### PR TITLE
CRM-19372 allow payment processors to define an array of accepted cre…

### PR DESCRIPTION
**Pull request by seamuslee001**
…dit card types

### CODE REVIEW

**Comment by seamuslee001** [comment]
ping @eileenmcnaughton @andrew-cormick-dockery

File: CRM/Financial/BAO/PaymentProcessor.php [hunk 1]
```diff
@@ -91,6 +91,22 @@ public function __construct() {
   }
 
   /**
+   * Retieve array of allowed credit cards for this payment processor.
+   * @param interger|null $paymentProcessorID id of processor.
```
Comment (andrew-cormick-dockery) [inline CRM/Financial/BAO/PaymentProcessor.php]: sp - Retrieve

**Review by jitendrapurohit — COMMENTED** [review]
I was able to contribute using VISA CC number eventhough I've not selected as a accepted CC cards for my payment processor. Is fix for this need to be included ? Steps to replicate -
- Create a dummy PP selecting only AMEX as an accepted credit card.
- Use this processor for a contribution page.
- Only Amex is shown on the live page (correct)
- Contribute using demo CC number for VISA (4111 1111 1111 1111).
- Contribution for the contact successfully gets completed. I think there should be validation error thrown on the first page ?

Thanks.

**Comment by seamuslee001** [comment]
Hi @jitendrapurohit I have added in some form validation and also fixed the foreach error you found

File: CRM/Core/Payment/Form.php [hunk 2]
```diff
@@ -323,11 +326,18 @@ public static function setDefaultValues(&$form, $contactID) {
    * Make sure that credit card number and cvv are valid.
    * Called within the scope of a QF formRule function
    *
+   * @pram int $processorID
```
Comment (jitendrapurohit) [inline CRM/Core/Payment/Form.php]: typo param**

File: CRM/Admin/Form/PaymentProcessor.php [hunk 3]
```diff
@@ -211,7 +211,9 @@ public function buildQuickForm($check = FALSE) {
     // is this processor active ?
     $this->add('checkbox', 'is_active', ts('Is this Payment Processor active?'));
     $this->add('checkbox', 'is_default', ts('Is this Payment Processor the default?'));
-
+    $creditCardTypes = CRM_Contribute_PseudoConstant::creditCard();
+    $this->addCheckBox('accept_credit_cards', ts('Select Credit Card Types that this payment processor can accept'),
```
Comment (jitendrapurohit) [inline CRM/Admin/Form/PaymentProcessor.php]: seems to be a long label for a field. Can we not shorten this to something `Accepted Credit Card Type(s)` or similar and move the current label to description ?

**Comment by seamuslee001** [comment]
thanks @jitendrapurohit have made that change and agree it makes more sense.

**Review by jitendrapurohit — APPROVED** [review]
Tested the updated changes and card types seem to be working fine for the processors. Also tested on upgrade(normal, multilingual). I hope we can merge this on build pass. Thanks for the quick update @seamuslee001

**Comment by jitendrapurohit** [comment]
Test failures are unrelated.
